blockindex/sgd_indexer.go

```diff
@@ -7,50 +7,474 @@ package blockindex
+}
+
+func (sgd *sgdRegistry) handleContractDisapproved(b batch.KVStoreBatch, log *action.Log) error {
+	var (
+		sgdIndex *indexpb.SGDIndex
+		err      error
+		event    struct {
+			ContractAddress common.Address
+		}
+	)
+	if err := _sgdABI.UnpackIntoInterface(&event, "ContractDisapproved", log.Data); err != nil {
+		return err
+	}
+
+	sgdIndex, err = sgd.getSGDIndex(event.ContractAddress.Bytes())
+	if err != nil {
+		return err
+	}
+	if !sgdIndex.Approved {
+		return errors.New("contract is not approved")
+	}
+	sgdIndex.Approved = false
+	return sgd.putIndex(b, sgdIndex)
+}
+
+func (sgd *sgdRegistry) handleContractRemoved(b batch.KVStoreBatch, log *action.Log) error {
+	var (
+		event struct {
+			ContractAddress common.Address
+		}
+	)
+	if err := _sgdABI.UnpackIntoInterface(&event, "ContractRemoved", log.Data); err != nil {
+		return err
+	}
+	return sgd.deleteIndex(b, event.ContractAddress.Bytes())
+}
+
+func (sgd *sgdRegistry) putIndex(b batch.KVStoreBatch, sgdIndex *indexpb.SGDIndex) error {
+	sgdIndexBytes, err := proto.Marshal(sgdIndex)
+	if err != nil {
+		return err
+	}
+	b.Put(_sgdBucket, sgdIndex.Contract, sgdIndexBytes, "failed to put sgd index")
+	return nil
+}
+
+func (sgd *sgdRegistry) deleteIndex(b batch.KVStoreBatch, contract []byte) error {
+	b.Delete(_sgdBucket, contract, "failed to delete sgd index")
 	return nil
 }
 
+// DeleteTipBlock deletes the tip block from SGDIndexer
+func (sgd *sgdRegistry) DeleteTipBlock(context.Context, *block.Block) error {
+	return errors.New("cannot remove block from indexer")
+}
+
+// CheckContract checks if the contract is a SGD contract
 func (sgd *sgdRegistry) CheckContract(ctx context.Context, contract string) (address.Address, uint64, bool, error) {
-	return nil, 0, false, nil
+	addr, err := address.FromString(contract)
+	if err != nil {
+		return nil, 0, false, err
+	}
+	sgdIndex, err := sgd.getSGDIndex(addr.Bytes())
+	if err != nil {
+		return nil, 0, false, err
+	}
+
+	addr, err = address.FromBytes(sgdIndex.Receiver)
+	return addr, _sgdPercentage, sgdIndex.Approved, err
+}
+
+// getSGDIndex returns the SGDIndex of the contract
+func (sgd *sgdRegistry) getSGDIndex(contract []byte) (*indexpb.SGDIndex, error) {
+	buf, err := sgd.kvStore.Get(_sgdBucket, contract)
+	if err != nil {
+		return nil, err
+	}
+	sgdIndex := &indexpb.SGDIndex{}
+	if err := proto.Unmarshal(buf, sgdIndex); err != nil {
+		return nil, err
+	}
+	return sgdIndex, nil
+}
+
+// FetchContracts returns all contracts that are eligible for SGD
+func (sgd *sgdRegistry) FetchContracts(ctx context.Context) ([]*SGDIndex, error) {
+	_, values, err := sgd.kvStore.Filter(_sgdBucket, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil {
+		if errors.Cause(err) == db.ErrNotExist || errors.Cause(err) == db.ErrBucketNotExist {
+			return nil, errors.Wrapf(state.ErrStateNotExist, "failed to get sgd states of ns = %x", _sgdBucket)
+		}
+		return nil, err
+	}
+	sgdIndexes := make([]*SGDIndex, 0, len(values))
```

Comment:
`sgdIndexPb := indexpb.SGDIndex{}`
move to here, can re-use the same var in the for loop below